File: AAAA_TOP200/岛屿最大面积.cpp



class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {

        int m = grid.size();
        int n = grid[0].size();

        int res =0;
        for(int i=0;i<m;i++) {
            for(int j=0;j<n;j++) {
                if(grid[i][j]==1) {
                    res++;
                    res = max(res,dfs(grid,i,j));
                }
            }
        }

        return res;
    }

    int dfs(vector<vector<int>> & grid, int i, int j) {
        int m = grid.size();
        int n = grid[0].size();

        if(!(i>=0&&i<m&&j>=0&&j<n&&grid[i][j]==1)) {
            return 0;
        }
        if(grid[i][j]!=1) {
            return 0;
        }

        grid[i][j]= 2;
        return 1+dfs(grid,i+1,j)
                +dfs(grid,i-1,j)
                +dfs(grid,i,j+1)
                +dfs(grid,i,j-1);

    }
};

File: AAAA_TOP100/岛屿最大面积.java



public class B {

    public static int maxAreaOfIsland(int[][] grid) {
        int ans = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                // 发现陆地！
                if (grid[i][j] == 1) {
                    ans = Math.max(ans, dfs(grid, i, j));
                }
            }
        }
        return ans;
    }

    static int dfs(int[][] grid, int m, int n) {

        if (!(m >= 0 && m < grid.length && n >= 0 && n < grid[0].length)) {
            return 0;
        }
        if (grid[m][n] == 0 || grid[m][n] == 2) {
            return 0;
        }
        grid[m][n] = 2;
        return 1+dfs(grid,m+1,n)
                +dfs(grid,m-1,n)
                +dfs(grid,m,n+1)
                +dfs(grid,m,n-1);
    }
    public static void main(String[] args) {

        int[][] grid = {{0,0,1,0,0,0,0,1,0,0,0,0,0},{0,0,0,0,0,0,0,1,1,1,0,0,0},{0,1,1,0,1,0,0,0,0,0,0,0,0},
                        {0,1,0,0,1,1,0,0,1,0,1,0,0},{0,1,0,0,1,1,0,0,1,1,1,0,0},{0,0,0,0,0,0,0,0,0,0,1,0,0},
                        {0,0,0,0,0,0,0,1,1,1,0,0,0},{0,0,0,0,0,0,0,1,1,0,0,0,0}};

        System.out.println(maxAreaOfIsland(grid));
    }
}